# STA130: Week 01 Homework

## Statistical Analysis of Titanic Passengers

This is a statistical analysis of a comma-separated-value file that contains data entries of passengers aboard the Titanic, including name, age, and sex.

Perhaps more interestingly, there is missing data within these entries found and summarized by **Chat-GPT**:

**Dataset Details**
- **Title:** Titanic Dataset
- **URL:** [Titanic Dataset CSV](https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv)
- **Description:** This dataset contains information about passengers aboard the Titanic. It includes various features such as age, sex, class, and survival status. The dataset provides insights into the demographics of the passengers and their survival outcomes.

Chat-GPT Log: https://chatgpt.com/share/da75cbaa-d8c4-48a2-9bb7-a6536077e228

**Note**: The logs for each GPT interaction will be listed at the top of each response in a markdown cell.
___

# 1. Dataset Selection

Here, we load our dataset and provide a total of missing values. We notice that most value types are not missing. while cabin ID and age have many missing values, as well as 2 missing embarked statuses.

Chat-GPT Log: https://chatgpt.com/share/6f7bbdf1-b6a1-43ac-a263-c6647b5562bc

In [48]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

# Check for missing values in the dataset
missing_values = td.isna().sum()

print(missing_values)

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64


## GPT Session Summary


___

# 2. Datset Dimensions

Chat-GPT Log: https://chatgpt.com/share/f88357c4-9dc7-4a07-b652-af3f2fa64787

In [49]:
# Get the dimensions (number of rows and columns)
dataset_shape = td.shape

# Get the column names
column_info = td.columns

print(f"Dataset Shape (Rows, Columns): {dataset_shape}")
print("\nColumn Names:")
print(column_info)

Dataset Shape (Rows, Columns): (891, 12)

Column Names:
Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')


## Observations and Variables

Here is how Chat-GPT described the observations and variables in the context of this dataset:

### 1. **Observations:**

- **Observations** refer to individual entries or records in the dataset, representing each **passenger** who was aboard the Titanic.
- Each row in the dataset corresponds to one passenger and contains various details about them, such as whether they survived, their age, gender, class, etc.
- For example, if there are 891 rows, that means there are 891 observations or passengers represented in this dataset.

### 2. **Variables:**

- **Variables** are the different types of data collected for each observation (passenger). These correspond to the **columns** in the dataset.
- In the Titanic dataset, the variables include:
    - **PassengerId**: A unique identifier for each passenger.
    - **Survived**: A binary variable where 1 indicates survival and 0 indicates that the passenger did not survive.
    - **Pclass**: The passenger class (1 = 1st class, 2 = 2nd class, 3 = 3rd class).
    - **Name**: The name of the passenger.
    - **Sex**: The gender of the passenger (male or female).
    - **Age**: The age of the passenger in years.
    - **SibSp**: The number of siblings or spouses aboard the Titanic with the passenger.
    - **Parch**: The number of parents or children aboard the Titanic with the passenger.
    - **Ticket**: The ticket number for the passenger.
    - **Fare**: The fare paid by the passenger.
    - **Cabin**: The cabin number, if available.
    - **Embarked**: The port of embarkation (C = Cherbourg; Q = Queenstown; S = Southampton).

In summary:

- **Observations** = individual passengers (rows).
- **Variables** = characteristics or details about those passengers (columns).

## General Definitions

Based on GPT's response, the following would be suitable definitions for statistical observations and varibales:

- **Observation:** a record of a measured object, often described by multiple variables and attributes; a row in a set of data.

- **Varibales:** a characteristic or attribute that describes an object and can take on different values, both qualitative and quantitative; the columns in a set of data.

## GPT Session Summary

___

# 3. Dataset Summary

Chat-GPT Log: https://chatgpt.com/share/f88357c4-9dc7-4a07-b652-af3f2fa64787

In [50]:
# General dataset info (column names, types, and missing values)
td.info()

# Summary statistics for numeric columns
print(td.describe())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB
       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    

In [51]:
# Show the first 20 rows of the dataset
td.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

Notice how the above uses 1s and 0s as boolean True and False values for survival status.

In [52]:
td['Survived'] = td['Survived'].astype(bool) # easily done because there are no missing survivor statuses
td.head(20)

PassengerId  Survived  Pclass  \
0             1     False       3   
1             2      True       1   
2             3      True       3   
3             4      True       1   
4             5     False       3   
5             6     False       3   
6             7     False       1   
7             8     False       3   
8             9      True       3   
9            10      True       2   
10           11      True       3   
11           12      True       1   
12           13     False       3   
13           14     False       3   
14           15     False       3   
15           16      True       2   
16           17     False       3   
17           18      True       2   
18           19     False       3   
19           20      True       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S  
15      0            248706  16.0000   NaN        S  
16      1            382652  29.1250   NaN        Q  
17      0            244373  13.0000   NaN        S  
18      0            345763  18.0000   NaN        S  
19      0              2649   7.2250   NaN        C

In [53]:
# Display value counts for categorical columns (like 'Survived' and 'Pclass')
print("\nSurvived counts:")
print(td['Survived'].value_counts())

print("\nPclass counts:")
print(td['Pclass'].value_counts())

print("\nSex counts:")
print(td['Sex'].value_counts())

print("\nEmbarked counts:")
print(td['Embarked'].value_counts())


Survived counts:
Survived
False    549
True     342
Name: count, dtype: int64

Pclass counts:
Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Sex counts:
Sex
male      577
female    314
Name: count, dtype: int64

Embarked counts:
Embarked
S    644
C    168
Q     77
Name: count, dtype: int64


Note, the embarked variable takes on 3 possible values:
- 'S' for Southampton
- 'C' for Cherbourg
- 'Q' for Queenstown

## GPT Session Summary




___

# 4. Missing Values in Dataset

Chat-GPT Log: https://chatgpt.com/share/f88357c4-9dc7-4a07-b652-af3f2fa64787

## Describe the discrepances between <code>td.shape</code> and <code>td.describe()</code> regarding:

- (a) the number of columns analyzed
- (b) the values reported in the "count" column

## (a)

Using <code>td.shape</code>, in Question 2, the dataset dimensions are (891, 12), meaning there are 891 rows (observations) and 12 columns (variables). However, <code>td.describe()</code> returns 7 columns: PassengerId, Survived, Pclass, Age, SibSp (sibling/spouses aboard), Parch (parents/children aboard), and Fare (price).

This occurs because these 7 columns contain numeric variables, and <code>td.describe()</code> only includes numeric columns, as it outputs statistical calculations like Standard Deviation in its descriptions. As seen above, the 7 returned columns are numeric, and the 5 columns left out (Name, Sex, Ticket, Cabin, Embarked) are all non-numeric. The sum of the numeric and non-numeric columns is 12, the number of columns returned by <code>td.shape</code>. 

So, <code>td.shape</code> returns the number of all columns, while <code>td.describe()</code> only returns numeric columns.

# (b)

Since there are 891 rows, it would be intuitve for each variable to have 891 values for each observation. While this is the case for variables like PassengerID or Name, there are 177 missing ages, 687 missing cabin numbers, and 2 missing embarked cities, as reported by <code>td.isna().sum()</code>.

So, while <code>td.shape</code> returns 891 rows, <code>td.describe()</code> only returns 714 values of ages. This discrepancy occurs simply because there are 177 rows that do not have an age listed for a given passenger. Instead, those missing values are represented by 'NaN', meaning not a number (recall: <code>td.describe()</code> only returns numeric values).

## GPT Session Summary

Here's a concise summary of our interaction:

---

In our discussion about the Titanic dataset:

(a) **Number of Columns Analyzed**:
- **td.shape** indicates the dataset has 12 columns (891 rows and 12 columns).
- **td.describe()** returns statistics for only 7 columns: PassengerId, Survived, Pclass, Age, SibSp, Parch, and Fare. This is because **td.describe()** only includes numeric columns, leaving out non-numeric columns (Name, Sex, Ticket, Cabin, Embarked).

(b) **Values in the "Count" Column**:
- **td.shape** suggests each of the 12 columns should have 891 values.
- **td.describe()** shows fewer values for columns with missing data. For example, Age has 714 values reported, indicating 177 missing entries. Missing values are represented as 'NaN' and are excluded from **td.describe()**'s output.

---

Let me know if you need any adjustments!
___

# 5. Attribute vs. Method

Chat-GPT Log: https://chatgpt.com/share/f88357c4-9dc7-4a07-b652-af3f2fa64787

## Attributes:

An attribute, like <code>td.shape</code>, is a **property of an object, holding some form of data** that can be accessed without performing any action on an object. An attribute does not use parantheses because it doesn't perform anything **on** the object, it simply gives details **about** the object. For example, <code>td.shape</code> provides the structure of the pandas DataFrame object by returning the number of rows and columns, without performing anything on the DataFrame.


## Method:

A method, like <code>td.describe()</code>, is a **function that performs an operation on an associated object, and may return a result of the object's data**. Since a method is a type of function (more specifically, a function associated with an object), they require parentheses when called (regardless of if they take arguments or not).

### Difference Between a Method and a Function
The main takeaway between a method and a function is that all methods are functions, but not all functions are methods; that is, a method is a specific type of function.

Functions are definitions of reusable code that aren't tied to any object, while a method has an associated object that it performs operations on.

### Difference Between a Function in Mathematics and Programming
Mathematical functions are well-defined in the sense that an input will **always** lead to the same output; semantically, we would call such a function **deterministic**. 

Programming functions **do not** guarantee that an input will **always** lead to the same output. One instance of this could be caused by "side effects," which could occur when passing a global variable into a function and then modifying that variable. There is also an element of uncertainty in programming functions, with programming languages being able to produce random integers that may be used to manipulate an input, leading to a random output (although, we **technically** cannot produce true randomness in computers).

## GPT Session Summary
Here’s a balanced summary:

- **Attributes**: Variables within a class or object that store data. For example, a `car` class might have attributes like `color` and `make`.

- **Methods**: Functions defined within a class that act on the attributes of instances of that class. For instance, a `drive` method might change the car’s state.

- **Functions**: General blocks of code that perform tasks. They include:
  - **Math Functions**:Mathematical Functions: These are relations that assign each input exactly one output. Examples include f(x) = x^2 or g(x) = sin(x). They describe relationships and operations within mathematical contexts.
  - **Programming Functions**: hese are blocks of code designed to perform specific tasks or operations. They can manipulate data, perform calculations, or control program flow.
___

# 6. <code>td.describe()</code> Summary Statistics Definitions

Chat-GPT Log: https://chatgpt.com/share/66e3b26e-0320-8008-b75c-c149f1715eb9

1. Count: The total number of non-null observations for each variable. In Statistics, this would represent the sample size of a dataset.

<br>

2. Mean: Known as the artithmetic mean, this statistic is the sum of all observations divided by the count ($n$). Given a variable $x$, its arithmetic mean ($\bar{x}$) is denoted by:

$$\large\bar{x} = \dfrac{1}{n}\displaystyle\sum_{i=1}^{n} x_i$$

<br>

3. std: Known as Standard Deviation ($s$), this statistic measures the "spread" of the data around the mean. It is the square root of the average of the squared deviations from the mean (the square root of the square ensures only positive results):

$$\large s = \sqrt{\dfrac{1}{n-1}\displaystyle\sum_{i=1}^{n} (x_i - \bar{x})^2}$$

<br>

4. Min: The minimum (smallest) value in a dataset:
$$\large Min = min(X_1, X_2, \cdots, X_n)$$

<br>

5. 25%: Known as the first quartile ($Q_1$), this statistic measures the value below which a quarter of the data points are found when arranged in increasing order. It measures the lower portion of the data's spread.

<br>

6. 50%: This represents the median ($Q_2$), or the value in the "middle" of the sorted dataset. So, half of the datapoints lie below this value, while the other half lie above this value.

<br>

7. 75%: The third quartile ($Q_3$) marks the value at which 75% of the datapoints lie below, meaning this value is the upper bound of three quarters if the data.

<br>

8. Max: The maximum (greatest) value observed in the dataset; the upper boundary of the set.

$$\large Max = max(X_1, X_2, \cdots, X_n)$$

## GPT Session Summary

Here’s a summary focusing on the definitions:

1. **Summary Statistics**:
   - **`count`**: Number of non-null observations.
   - **`mean`**: Average value of the data.
   - **`std`**: Standard deviation, measuring the dispersion of the data.
   - **`min`**: Minimum value in the data.
   - **`25%`**: 25th percentile (first quartile), the value below which 25% of the data fall.
   - **`50%`**: Median (50th percentile), the middle value of the data.
   - **`75%`**: 75th percentile (third quartile), the value below which 75% of the data fall.
   - **`max`**: Maximum value in the data.

___

# 7. Missing Data: <code>td.dropna()</code> and <code>del td['col']</code>

Chat-GPT Log: https://chatgpt.com/share/66e3b26e-0320-8008-b75c-c149f1715eb9

<code>td.dropna()</code> deletes an entire row of data if a single data point within that row is missing a variable; that is, if any of the associated columns fo a row are missing. By default, the axis of this method is 0, but changing that axis to 1 will cause the method to delete columns of data given an associated row that is null: 
<code>td.dropna(axis=1)</code>.

On the other hand, <code>del td['col']</code> deletes an entire column of data upon request.
 
1. Using <code>td.dropna()</code> would be preffered over <code>del td['col']</code> in a case where there is more missing data in a given row rather than a given column. This preserves as much valuable data as possible by only deleting the data which has missing values associated with it. For example, if a row of data was missing a passenger's name, it wouldn't make sense to delete the entire column of passenger names because of one missing data point. Instead, we would use <code>td.dropna()</code> to drop the row with the missing name.

<br>

2. Contrarily to 7.1, <code>del td['col']</code> would be useful if there were more data missing in a given column than a row. This way, the unuseful data would be deleted along a column without affecting the rest of the data in a row. For example, in the titanic dataset, there are 687/891 cabin IDs missing. It wouldn't make sense to delete 687 rows of data to account for this. Instead, it would be wiser to delete the column of cabin IDs entirely and prevent any further valuable data from being deleted across a row. <br><br> This is further warranted if the column in question is not useful. The cabin ID column yields no statistical significance and any associated rows should not be deleted because of it.

<br>

3. In order to prevent as much unecessary data loss as possible, using <code>del td['col']</code> first would help <code>td.dropna()</code> target useful null values. In a similar premise to 7.2, it would make sense to delete the cabin IDs first. That way, there wouldn't be as many rows deleted from null values. For example, if there was just two rows, each with a missing value somewhere, deleting the cabin column first would not only prevent over 680 rows from being deleted, but would then allow for <code>td.dropna()</code> to target the rest of the null rows. Also, as stated in 7.2, insignificant columns like the cabin ID can be deleted without consequence because there isn't much stastical analysis you can perform on such data, and would allow for focusing <code>td.dropna()</code> on relevant data.

<br>

4. As explored earlier, it would make sense to first delete the Cabin column, as there is no statistical significance to this column, and just 204 non-null values would not be a big enough sample size to work with anyway. This prevents <code>td.dropna()</code> from deleting 687 rows of useful data; this would delete 177 rows of missing ages and 2 rows of embarked status.<br>**Note**: Deleting the ages column would not be smart as age is a useful variable for statistical analysis, and there isn't a relatively large amount of data missing.

In [66]:
dropped_td = td.copy() # to compare before and after

if ('Cabin' in td.columns):
    del dropped_td['Cabin']

dropped_td = dropped_td.dropna()

dropped_td.head(20)

survived  pclass     sex   age  sibsp  parch      fare embarked   class  \
1           1       1  female  38.0      1      0   71.2833        C   First   
3           1       1  female  35.0      1      0   53.1000        S   First   
6           0       1    male  54.0      0      0   51.8625        S   First   
10          1       3  female   4.0      1      1   16.7000        S   Third   
11          1       1  female  58.0      0      0   26.5500        S   First   
21          1       2    male  34.0      0      0   13.0000        S  Second   
23          1       1    male  28.0      0      0   35.5000        S   First   
27          0       1    male  19.0      3      2  263.0000        S   First   
52          1       1  female  49.0      1      0   76.7292        C   First   
54          0       1    male  65.0      0      1   61.9792        C   First   
62          0       1    male  45.0      1      0   83.4750        S   First   
66          1       2  female  29.0      0      0   10.5000        S  Second   
75          0       3    male  25.0      0      0    7.6500        S   Third   
88          1       1  female  23.0      3      2  263.0000        S   First   
92          0       1    male  46.0      1      0   61.1750        S   First   
96          0       1    male  71.0      0      0   34.6542        C   First   
97          1       1    male  23.0      0      1   63.3583        C   First   
102         0       1    male  21.0      0      1   77.2875        S   First   
110         0       1    male  47.0      0      0   52.0000        S   First   
118         0       1    male  24.0      0      1  247.5208        C   First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
6      man        True    E  Southampton    no   True  
10   child       False    G  Southampton   yes  False  
11   woman       False    C  Southampton   yes   True  
21     man        True    D  Southampton   yes   True  
23     man        True    A  Southampton   yes   True  
27     man        True    C  Southampton    no  False  
52   woman       False    D    Cherbourg   yes  False  
54     man        True    B    Cherbourg    no  False  
62     man        True    C  Southampton    no  False  
66   woman       False    F  Southampton   yes   True  
75     man        True    F  Southampton    no   True  
88   woman       False    C  Southampton   yes  False  
92     man        True    E  Southampton    no  False  
96     man        True    A    Cherbourg    no   True  
97     man        True    D    Cherbourg   yes  False  
102    man        True    D  Southampton    no  False  
110    man        True    C  Southampton    no   True  
118    man        True    B    Cherbourg    no  False

In [55]:
dropped_td.describe()

PassengerId      Pclass         Age       SibSp       Parch        Fare
count   712.000000  712.000000  712.000000  712.000000  712.000000  712.000000
mean    448.589888    2.240169   29.622191    0.514045    0.432584   34.567251
std     258.683191    0.836854   14.502891    0.930692    0.854181   52.938648
min       1.000000    1.000000    0.000000    0.000000    0.000000    0.000000
25%     222.750000    1.000000   20.000000    0.000000    0.000000    8.050000
50%     445.000000    2.000000   28.000000    0.000000    0.000000   15.645850
75%     677.250000    3.000000   38.000000    1.000000    1.000000   33.000000
max     891.000000    3.000000   80.000000    5.000000    6.000000  512.329200

In [56]:
print(dropped_td.isna().sum()) # Ensure no null values

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64


## GPT Session Summary

1. **`dropna()` Function**:
   - The `dropna()` function removes missing values from a DataFrame or Series. It can drop rows or columns with NaNs, and its parameters include `axis` (rows/columns), `how` (criteria for dropping), `thresh` (minimum number of non-NaN values), `subset` (specific columns to consider), and `inplace` (whether to modify the original DataFrame).

2. **`del df['col']` vs. `dropna()`**:
   - `del df['col']` removes a specific column from the DataFrame, while `dropna()` removes rows or columns with NaNs. Using `dropna()` might be preferred if you want to retain as much data as possible by removing only incomplete rows or columns. 

3. **Use Case for `del df['col']` Before and With `dropna()`**:
   - Deleting irrelevant columns before using `dropna()` helps prevent unnecessary removal of rows. For example, removing columns with high proportions of missing data first ensures `dropna()` only affects relevant columns, thereby preserving valuable data and improving data processing efficiency.


This summary should provide a clear overview of the key points discussed.

___ 

# 8. GPT Experimentation

## 1. Group by Columns

Chat-GPT Log: https://chatgpt.com/share/66e3b22b-611c-8008-986d-4f1d63c51224

In [57]:
import pandas as pd

url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

new_td = pd.read_csv(url)

new_td

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

### GPT's Response:

The content of the discussion covers the use of the pandas function `td.groupby("col1")["col2"].describe()`, which performs group-wise statistical analysis in a dataset. This function is essential for understanding variations in data across different categories by grouping the data based on the unique values in one column (`"col1"`) and then calculating a set of descriptive statistics (such as count, mean, standard deviation, min, max, and percentiles) for another column (`"col2"`) within each group.

### Example in the Titanic dataset:
By using the command `td.groupby("sex")["age"].describe()`, the dataset is grouped by the `"sex"` column, and the `"age"` column is analyzed within each group. This provides a detailed breakdown of the age distribution for male and female passengers, including:
- **Count**: Number of non-missing age values in each group.
- **Mean**: Average age for males and females.
- **Standard Deviation**: How much ages deviate from the mean in each group.
- **Percentiles**: 25th, 50th (median), and 75th percentiles, showing how ages are distributed across the group.
- **Min and Max**: The youngest and oldest ages for each gender.

This type of analysis is critical for revealing patterns or differences in data, such as age distribution by gender, which may have implications in various contexts, such as survival rates or passenger demographics.

In [58]:
new_td.groupby("sex")["age"].describe()

count       mean        std   min   25%   50%   75%   max
sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0

### Additional Example:

Use <code>new_td.groupby('pclass')['age'].describe()</code> to analyze the ages of passengers by their class.

In [59]:
new_td.groupby('pclass')['age'].describe()

count       mean        std   min   25%   50%   75%   max
pclass                                                           
1       186.0  38.233441  14.802856  0.92  27.0  37.0  49.0  80.0
2       173.0  29.877630  14.001077  0.67  23.0  29.0  36.0  70.0
3       355.0  25.140620  12.495398  0.42  18.0  24.0  32.0  74.0

It is clear that, on average, higher-class passengers (ie. first class) are older than lower class passengers. This makes sense since older people are generally more affluent and have greater status.

## 2. Analyzing "Missing" Values
*No GPT used*

In [60]:
new_td.describe()

survived      pclass         age       sibsp       parch        fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

At first glance, it might seem as if there were many missing counts in <code>new_td.groupby('pclass')['age'].describe()</code>, but the above method is not counting the total number of passengers; it is counting the total number of passengers **by class**. That is, there are 186 first class passengers, 173 second class passengers, and 355 third class passengers, summing to 714 passengers. The 177 missing passengers are just rows with null values for class.

So, we can generalize this by saying the count of <code>new_td.groupby("col1")["col2"].describe()</code> counts the number of entries, sorted by 'col1', and then describes those sorted groups in the context of 'col2.'

# 3. GPT vs. Google: 7 Game Series  
*(a more exciting matchup than last year's Finals)*

Chat-GPT Log: https://chatgpt.com/share/66e3b1ef-6c64-8008-a37c-19c09e871df3

## (a) Forget to include <code>import pandas as pd</code> in your code

In [61]:
# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

### GPT-Response:
It looks like you're running into an error where `pd` is not defined. This usually means that you need to import the pandas library, which is typically imported with the alias `pd`. You can fix this by adding the following import statement at the beginning of your code:

```python
import pandas as pd
```

If you have already imported pandas and are still seeing the error, make sure that the import statement is placed before the code where you use `pd`.

### Google Search
The solution to my problem was found relatively quickly via a [Stack Overflow Thread](https://stackoverflow.com/a/68904453)

### Comparison
While both solutions were found relatively quickly, Chat-GPT's response was easier to access. Rather than looking through forums for the answer myself, it took less than 5 seconds within providing a prompt to find my answer.

## (b) Mistype "titanic.csv" as "titanics.csv"

In [62]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

### GPT-Response:
It looks like you're running into an error where `pd` is not defined. This usually means that you need to import the pandas library, which is typically imported with the alias `pd`. You can fix this by adding the following import statement at the beginning of your code:

```python
import pandas as pd
```

If you have already imported pandas and are still seeing the error, make sure that the import statement is placed before the code where you use `pd`.

### Google Search:

After searching for about 3 minutes, intentionally tweaking my queries to hint at the error, I finally caught a somewhat similar issue that was resolved in a [Python Forum](https://python-forum.io/thread-24424.html). In this case, a user accidentally put 4 W's when typing www.google.com. No solution to my specific case was found

### Comparison:

GPT easily wins this round, as it not only answered my specific problem, but did so within seconds. Searching with Google took me over 3 minutes and did not even provide a solution specific to my case.

### Bonus: Experimenting With Other Typos
SyntaxErrors, like removing an equal sign assignment operator, is easily caught by the Python console and the error is quickly pointed to for the user to fix.
Even AttributeErrors, like misspelling read_csv, are easily caught and fairly simple for the user to find the mistake.

## (c) Using a Dataframe Without Variable Assignment
*e.g. Misspelling a variable name when calling it*

In [63]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

Td.describe()

NameError: name 'Td' is not defined

### GPT-Response:
It looks like there’s a case sensitivity issue in your code. In Python, variable names are case-sensitive, so `td` and `Td` are considered different variables.

You should use the same case for your variable names consistently. Since you defined your DataFrame as `td`, you should use `td` (all lowercase) in your `describe()` method call:

```python
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

# Display the summary statistics
td.describe()
```

This should fix the `NameError`. Make sure to use the correct variable name throughout your code.

### Google Search:

After about a minute of searching, I found an article about solutions for NameErrors in Python:
https://careerkarma.com/blog/python-nameerror-name-is-not-defined/. 

The first solution highlighted typos in code when calling a variable that seems to be defined, which is the case of the issue above.


### Comparison:

GPT is again the winner here, as it answers my question quickly and specifically to my case. Additionally, knowing the error I had planted creates bias when looking up Google searches as I knew which articles to look for. In an actual case of debugging, I might have completely missed the article I cited.

## (d) Missing Parantheses

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url

Td.describe()

### GPT-Response:
It looks like there's a syntax error due to a missing closing parenthesis in your `pd.read_csv()` function call. Make sure you close all parentheses properly. Here's the corrected code:

```python
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

# Display the summary statistics
td.describe()
```

Ensure that each opening parenthesis `(` has a corresponding closing parenthesis `)`.

### Google Search:
While the Google Search only took me about 2 minutes, this was still far longer than I expected. Perhaps, such a simple error is not reported as much and so will not be fed to the top of the search results of Google. Solution:
https://stackoverflow.com/a/33282998

### Comparison:
While I thought this would be a close call, GPT wins again. Having access and memory to my code makes it much easier for the chat-bot to identify my error. Maybe it is picking up my errors easily because I am doing all of Q.8 in one session, and so the bot has greater memory to make conclusions.

## (e) Typo in Chained Functions

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

td.groupby("survived")["age"].decribe()

### GPT-Response:
It seems like there might be a typo in your code. The error message suggests that you're trying to call `describe()` on a `SeriesGroupBy` object, which is likely due to a misspelling of `describe()`.

Here's the correct code to call the `describe()` method on a DataFrame:

```python
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

# Display the summary statistics
td.describe()
```

If you were trying to use `describe()` on a `GroupBy` object, make sure that you’re applying `describe()` directly on a DataFrame or Series, not on the result of a group operation.

For example, if you were grouping by a column and then trying to call `describe()`, it should look something like this:

```python
grouped = td.groupby('class')
grouped.describe()  # This works on a DataFrameGroupBy object
```

Ensure that you’ve spelled `describe` correctly and that you're applying it to the right type of object.

### Google Search:
I could not find a solution on Google within a reasonable amount of time. This makes sense since this is such a specific, niche error that is hard to pinpoint without having the ability to process the code.

### Comparison:
Obviously, GPT is the winner here, is it not only found a solution quickly, but was able to actually spot the error, whereas my problem was not resolved by Google.

## (f) Typo in Columns

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

td.groupby("Survived")["age"].describe()

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

td.groupby("survived")["Age"].describe()

### GPT-Response:

#### Pt. 1 ("Survived")
It looks like you’re encountering an `HTTPError` while trying to read the dataset from the provided URL. Here are a few things to check and try:

1. **Verify the URL:** Double-check that the URL is correct. The URL you provided seems to be for a different Titanic dataset. The correct URL for the Seaborn Titanic dataset is:
   ```python
   url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
   ```

2. **Check URL Accessibility:** Ensure that you can access the URL from your browser or use a tool like `curl` or `wget` to confirm it's reachable.

3. **Try an Alternative Method:** If you have the dataset file locally or in another location, you can try loading it from that source. For example:
   ```python
   td = pd.read_csv("path/to/local/titanic.csv")
   ```

Here’s how you can update your code with the correct URL:

```python
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)
```

Give that a try and see if it resolves the issue!

#### Pt. 2 ("Age")
1. **Verify Exact Column Names:** Ensure you're using the correct capitalization for column names by printing them:
   
   ```python
   print(td.columns)
   ```

2. **Standardize Column Names:** Convert all column names to lowercase (or uppercase) to avoid capitalization issues:

   ```python
   td.columns = td.columns.str.lower()  # Convert all column names to lowercase
   ```

3. **Use Correct Column Name:** After standardizing, use the correct column name in your analysis. For example, if you standardized to lowercase:

   ```python
   grouped_description = td.groupby("survived")["age"].describe()
   print(grouped_description)
   ```

By ensuring consistency in capitalization, you can avoid errors related to column names.

### Google Search:
The best solution I could find was a Stack Overflow forum about a KeyError involving accidental whitespace in their code, which does not address my problem directly, but might give me an idea of a solution:
https://stackoverflow.com/questions/61204694/keyerror-occurs-even-though-column-exists

### Comparison:
An obvious trend has been formed where GPT having full access and memory to my specific problem and code make it easier to spot errors in my program. Once again, GPT found the solution while Google could not in a reasonable amount of time.

## (g) Typo in Columns

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

td.groupby(survived)["age"].describe()

In [ ]:
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

td.groupby("survived")[age].describe()

### GPT-Response:

#### Pt. 1 (Survived)
The `NameError` indicates that `survived` is being interpreted as a variable rather than a string. Column names should be enclosed in quotes. Make sure you use `"survived"` (in quotes) as the column name. Here’s the corrected code:

```python
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

# Strip spaces from column names
td.columns = td.columns.str.strip()

# Group by 'survived' and describe the 'age' column
grouped_description = td.groupby("survived")["age"].describe()

print(grouped_description)
```

Ensure that column names are used as strings, enclosed in quotes, to avoid such errors.

#### Pt. 2 (Age)
It looks like you’re missing quotes around the `"age"` column name. Column names should always be provided as strings. Here’s the corrected code:

```python
import pandas as pd

# URL of the Titanic dataset
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

# Load the dataset
td = pd.read_csv(url)

# Strip spaces from column names
td.columns = td.columns.str.strip()

# Group by 'survived' and describe the 'age' column
grouped_description = td.groupby("survived")["age"].describe()

print(grouped_description)
```

Make sure to use `"age"` (with quotes) instead of just `age`. This should resolve the `NameError`.

### Google Search:
Again, Google was not able to solve this problem. My Google searches yielded a combination of scope errors and python version errors, but nothing pertinent to my situation.

### Comparison:
Simply put: GPT Wins. It provided a coherent solution, while Google did not provide any solution.

## FINAL VERDICT
GPT (7) - (0) GOOGLE <br>
GPT Wins!

It's clear that Chat-GPT having access to my code and dataset makes it the superior debugger. Additionally, it is able to have a dialogue and ask the user follow-up questions to better target the problem. Chat-GPT is also more beneficial in the sense that the answer is brought **to you**, rather than you having to parse through search results.
___

## GPT Session Summary

1. **Issue Identified:**
   - Encountered a `NameError` due to `pd` not being defined. Solution: Import pandas with `import pandas as pd`.
   - `HTTPError` when trying to read the dataset from an incorrect URL. Solution: Use the correct URL for the Seaborn Titanic dataset.
   - Error due to a missing parenthesis in the `pd.read_csv()` function call.
   - `AttributeError` occurred because `describe()` was misspelled as `decribe`.
   - `KeyError` was encountered for column names. Solution: Check column names and ensure correct capitalization and spacing.
   - Additional `KeyError` and `NameError` occurred due to incorrect column name usage and missing quotes around column names.

2. **Solutions Provided:**
   - Verify and correct the dataset URL.
   - Ensure correct spelling and capitalization in method calls and column names.
   - Use correct string formatting for column names in DataFrame operations.

3. **Final Corrected Code:**
   ```python
   import pandas as pd

   # URL of the Titanic dataset
   url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"

   # Load the dataset
   td = pd.read_csv(url)

   # Strip spaces from column names
   td.columns = td.columns.str.strip()

   # Group by 'survived' and describe the 'age' column
   grouped_description = td.groupby("survived")["age"].describe()

   print(grouped_description)
   ```

# 9. Resources Used

Before accessing this homework, I did breifly look over the course wiki-textbook. However, I used the first lecture as a guide to using pandas and statistical analysis at a later point. When I did use ChatBots, it was ChatGPT; logs and summaries were included throughout the homework. I did set up the NotebookLLM ChatBot, but did not use it much.